[cpu] add VecConvert between 8bits and 16bits

Valentine233 opened the pull request:
The perf benefit was found in https://github.com/pytorch/pytorch/issues/124697#issuecomment-2071658300.

The PR adds intrinsic specializations between int8/uint8 and bf16/fp16.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

Valentine233 committed:
[cpu] add VecConvert between 8bits and 16bits

Valentine233 commented on aten/src/ATen/cpu/vec/vec256/vec256_convert.h:
```diff
@@ -128,30 +128,102 @@ struct VecConvert<int32_t, 1, uint8_t, 1> {
+  }
+};
+
+template <typename dst_t>
```
Just some clang-format modifications for conversions between 8bits and float.

Valentine233 commented on aten/src/ATen/cpu/vec/vec512/vec512_convert.h:
```diff
@@ -119,30 +119,102 @@ struct VecConvert<int32_t, 1, uint8_t, 1> {
+  }
+};
+
+template <typename dst_t>
```
Just some clang-format modifications for conversions between 8bits and float.

jgong5 commented on aten/src/ATen/cpu/vec/vec256/vec256_convert.h:
```diff
@@ -128,30 +128,102 @@ struct VecConvert<int32_t, 1, uint8_t, 1> {
   static inline VectorizedN<dst_t, 1> apply(
-      const VectorizedN<float, 1>& src) {
-    return convert_float_to_int8<dst_t>(src[0]);
+      const VectorizedN<BFloat16, 1>& src) {
+    __m256 value;
+    cvtbf16_fp32(_mm256_castsi256_si128(src[0]), value);
+    return convert_float_to_int8<dst_t>(value);
```
why not calling the existing convert from int8 to fp32 and fp32 to bf16/fp16? Then you can combine bf16 and fp16 into a single implementation.

Valentine233 replied:
Thanks for your advice. Modified it!

Valentine233 committed:
updatewq VecConvert between 8bits and 16bits

jgong5 reviewed (changes requested):
Why do you change vec_base.h?

Valentine233 committed:
update VecConvert between 8bits and 16bits

Valentine233 committed:
update VecConvert between 8bits and 16bits

Valentine233 commented:
> Why do you change vec_base.h?

Just added a function `is_bit8` in `vec_base.h` to simplify codes. Sorry for not realizing that `clang-format` brought such a big code change. I have removed the irrelevant change now, thanks.

jgong5 commented on aten/src/ATen/cpu/vec/vec_base.h:
```diff
@@ -90,6 +90,16 @@ struct is_reduced_floating_point:
 constexpr bool is_reduced_floating_point_v = is_reduced_floating_point<T>::value;
 
+template <typename T>
+struct is_bit8:
```
```suggestion
struct is_8bit_integer:
```

jgong5 commented on aten/src/ATen/cpu/vec/vec256/vec256_convert.h:
```diff
@@ -126,32 +126,48 @@ struct VecConvert<int32_t, 1, uint8_t, 1> {
+    float,
+    1,
+    typename std::enable_if_t<
+        std::is_same_v<dst_t, unsigned char> ||
+            std::is_same_v<dst_t, signed char>,
```
do you want to simplify similar checks like this too?

Valentine233 replied:
Sure!

Valentine233 committed:
update VecConvert between 8bits and 16bits

Valentine233 commented:
@pytorchbot merge
